### PR TITLE
RavenDB-15596 ETL fixes

### PR DESCRIPTION
**Pull request by karmeli87**
- process ETL items by the etag order
- Break large time series ETL into several batches
- Filter only after extraction, otherwise we might move the etag too far and skip items

### CODE REVIEW

File: src/Raven.Server/Documents/ETL/EtlProcess.cs [hunk 1]
```diff
@@ -100,6 +99,12 @@ public static EtlProcessState GetProcessState(DocumentDatabase database, string
         }
     }
 
+    public interface IExtractEnumerator<out T> : IEnumerator<T> where T : ExtractedItem
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ETL/EtlProcess.cs]: Let's extract it to separate file.

File: src/Raven.Server/Documents/ETL/EtlProcess.cs [hunk 2]
```diff
@@ -286,34 +262,27 @@ private IEnumerable<TExtracted> Extract(DocumentsOperationContext context, long
             EtlStatsScope stats,
             DisposableScope scope)
         {
-            var lastProcessedDocEtag = stats.GetLastTransformedOrFilteredEtag(EtlItemType.Document);
-            var toEtag = lastProcessedDocEtag > 0 
-                ? lastProcessedDocEtag 
-                // If there was no document transformed in current batch
-                // it means we're done with documents, we can send all time series
-                : long.MaxValue;
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ETL/EtlProcess.cs:294]: Just to make sure. We no longer need that because after your changes docs and timeseries are processed in etag order (not as it was before that we were processing docs and their time series were processed later on). Is that correct?

File: src/Raven.Server/Documents/ETL/Providers/Raven/Enumerators/AttachmentTombstonesToRavenEtlItems.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            _context = context;
+            _tombstones = tombstones;
+            _collections = collections ?? throw new ArgumentNullException(nameof(collections));
+        }
+
+        public bool Filter()
+        {
+            var tombstone = _tombstones.Current;
+            if (tombstone.Type != Tombstone.TombstoneType.Attachment)
+                throw new InvalidOperationException();
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ETL/Providers/Raven/Enumerators/AttachmentTombstonesToRavenEtlItems.cs]: Please provide exception message. Also you might extract the exception to a dedicated method to get slightly better perf. `Filter` is expected to be called quite frequently.

File: src/Raven.Server/Documents/ETL/EtlProcess.cs [hunk 4]
```diff
@@ -502,8 +455,7 @@ private bool CanContinueBatch(EtlStatsScope stats, TExtracted currentItem, int b
                 return false;
             }
 
-            if (currentItem.Type == EtlItemType.Document &&
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ETL/EtlProcess.cs:505]: Why did you remove `currentItem.Type == EtlItemType.Document` check? We should check `Database.Configuration.Etl.MaxNumberOfExtractedDocuments` only for the docs

File: src/Raven.Server/Documents/ETL/Providers/Raven/Enumerators/AttachmentTombstonesToRavenEtlItems.cs (hunk 3 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/ETL/Providers/Raven/Enumerators/AttachmentTombstonesToRavenEtlItems.cs]: And fix the missing exception message in other places as well.

File: src/Raven.Server/Documents/ETL/EtlProcess.cs (hunk 2 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/ETL/EtlProcess.cs:294]: yes

File: src/Raven.Server/Documents/ETL/Providers/Raven/Enumerators/AttachmentTombstonesToRavenEtlItems.cs (hunk 3 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/ETL/Providers/Raven/Enumerators/AttachmentTombstonesToRavenEtlItems.cs]: Frequent calls was my concern as well, but it is not suppose to be super performant anyway, right?
`Filter` will be called as much as `MoveNext`.

File: src/Raven.Server/Documents/ETL/EtlProcess.cs (hunk 4 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/ETL/EtlProcess.cs:505]: we do, one line below.

File: src/Raven.Server/Documents/ETL/Providers/Raven/Enumerators/AttachmentTombstonesToRavenEtlItems.cs (hunk 3 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/ETL/Providers/Raven/Enumerators/AttachmentTombstonesToRavenEtlItems.cs]: Yeah. Just moving the `throw` clause to a separate method and calling it here might let it optimize stuff by inlining.

File: src/Raven.Server/Documents/ETL/EtlProcess.cs (hunk 4 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/ETL/EtlProcess.cs:505]: Okay. Makes sense, you don't have to check that `currentItem` is a document

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Documents/ETL/EtlProcess.cs [hunk 5]
```diff
@@ -201,81 +194,61 @@ private IEnumerable<TExtracted> Extract(DocumentsOperationContext context, long
...
-            {
-                var attachmentTombstones = Database.DocumentsStorage
-                    .GetTombstonesFrom(context, AttachmentsStorage.AttachmentsTombstones, fromEtag, 0, long.MaxValue).GetEnumerator();
-                scope.EnsureDispose(attachmentTombstones);
-
-                attachmentTombstones = new FilterTombstonesEnumerator(attachmentTombstones, stats, Tombstone.TombstoneType.Attachment, context,
-                    fromCollections: Transformation.ApplyToAllDocuments ? null : Transformation.Collections);
 
-                merged.AddEnumerator(ConvertTombstonesEnumerator(attachmentTombstones, null, EtlItemType.Document));
+                if (ShouldTrackAttachmentTombstones())
```
Comment (haludi) [inline src/Raven.Server/Documents/ETL/EtlProcess.cs:223]: You should do check and track attachments in `Transformation.ApplyToAllDocuments` as well, no?

File: src/Raven.Server/Documents/ETL/EtlProcess.cs [hunk 6]
```diff
@@ -706,29 +654,26 @@ public void Run()
...
-                                    List<TTransformed> ExtractTransform(EtlItemType type)
-                                    {
-                                        var extracted = Extract(context, loadLastProcessedEtag + 1, type, stats);
-                                        return Transform(extracted, context, stats, state);
-                                    }
+                                    var transformations = Transform(merged, context, stats, state);
 
-                                    var successfulLoad = false;
-
-                                    if (transformations.Count > 0)
```
Comment (haludi) [inline src/Raven.Server/Documents/ETL/EtlProcess.cs:726]: In a case, there is no transformed item you still act as "didWork"

Comment (karmeli87) [reply src/Raven.Server/Documents/ETL/EtlProcess.cs:726]: yes, because there might be stuff that we skipped, but we need to update the etag.

File: src/Raven.Server/Documents/ETL/EtlProcess.cs [hunk 7]
```diff
@@ -431,7 +403,7 @@ public bool Load(IEnumerable<TTransformed> items, DocumentsOperationContext cont
 
                     stats.RecordLastLoadedEtag(stats.LastTransformedEtags.Values.Max());
 
-                    Statistics.LoadSuccess(stats.NumberOfTransformedItems.Sum(x => x.Value));
+                    Statistics.LoadSuccess(count);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ETL/EtlProcess.cs:406]: It's for stats only but I don't think that the usage of `count` is correct. `LoadInternal` returns the number of commands (for example we might have DELETE and PUT for a single transformed item) while we should record `transformed - filtered` items here

Comment (arekpalinski) [reply src/Raven.Server/Documents/ETL/EtlProcess.cs:406]: and we should use the same value for below `stats.RecordLoadSuccess()`

Comment (karmeli87) [reply src/Raven.Server/Documents/ETL/EtlProcess.cs:406]: To me it is not well defined.

does sending a tombstone counts as `Load` ? (it doesn't count as `Transformed`, it counts as `NumberOfTransformedTombstones`)
Load of a time-series is segment based, entries based or is it count as 1 regardless to the number of segments (`TransformedItems` is segment based btw.)
Same for counters.

Not sure that `count` is the best choice though..

Comment (arekpalinski) [reply src/Raven.Server/Documents/ETL/EtlProcess.cs:406]: As we discussed there are other problems with using the transformed number of items - for example we might transform time series segment but filter out all it's entries under the covers.

So the usage of `count` had some drawbacks but appears to be the best number for stats purpose.

**Review by arekpalinski — APPROVED** [review]

File: src/Raven.Server/Documents/ETL/Providers/Raven/Enumerators/CountersToRavenEtlItems.cs [hunk 8]
```diff
@@ -1,26 +1,39 @@
...
             _counters = counters;
             _collection = collection;
         }
 
+        private bool Filter(RavenEtlItem item)
+        {
+            var doc = _context.DocumentDatabase.DocumentsStorage.Get(_context, item.DocumentId, DocumentFields.Default);
+
+            // counter has lower etag than its document - we skip it to avoid
+            // sending a counter of document that can not exist on the destination side
```
Comment (ayende) [inline src/Raven.Server/Documents/ETL/Providers/Raven/Enumerators/CountersToRavenEtlItems.cs:26]: what will send the counter to the other side afterward? won't this cause us to miss a write?

**Review by ayende — COMMENTED** [review]
LGTM

**Review by ayende — APPROVED** [review]

File: src/Raven.Server/Documents/ETL/Providers/Raven/Enumerators/CountersToRavenEtlItems.cs (hunk 8 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/ETL/Providers/Raven/Enumerators/CountersToRavenEtlItems.cs:26]: we always load the counters together with the document itself.
https://github.com/ravendb/ravendb/blob/4569e8a7bf34cc94c722b85bd16e6f342d3df663/src/Raven.Server/Documents/ETL/Providers/Raven/RavenEtlDocumentTransformer.cs#L217
